sycl/source/detail/device_filter.cpp

```diff
@@ -42,6 +43,165 @@ std::vector<std::string_view> tokenize(const std::string &Filter,
   return Tokens;
 }
 
+// ---------------------------------------
+// ONEAPI_DEVICE_SELECTOR support
+
+static backend Parse_ODS_Backend(const std::string_view &BackendStr,
+                                 const std::string_view &FullEntry) {
+  // Check if the first entry matches with a known backend type
+  auto SyclBeMap =
+      getSyclBeMap(); // <-- std::array<std::pair<std::string, backend>>
+                      // [{"level_zero", backend::level_zero}, {"*", ::all}, ...
+  auto It = std::find_if(
+      std::begin(SyclBeMap), std::end(SyclBeMap), [&](auto BePair) {
+        return std::string::npos != BackendStr.find(BePair.first);
+      });
+
+  if (It == SyclBeMap.end()) {
+    // backend is required
+    std::stringstream ss;
+    ss << "ONEAPI_DEVICE_SELECTOR parsing error. Backend is required but "
+          "missing from \""
+       << FullEntry << "\"";
+    throw sycl::exception(sycl::make_error_code(errc::invalid), ss.str());
+  } else {
+    return It->second;
+  }
+}
+
+static void Parse_ODS_Device(ods_target &Target,
+                             const std::string_view &DeviceStr) {
+  // DeviceStr will be: 'gpu', '*', '0', '0.1', 'gpu.*', '0.*', or 'gpu.2', etc.
+  std::vector<std::string_view> DeviceSubPair = tokenize(DeviceStr, ".");
+  std::string_view TopDeviceStr = DeviceSubPair[0];
+
+  // Handle explicit device type (e.g. 'gpu').
+  auto DeviceTypeMap =
+      getSyclDeviceTypeMap(); // <-- std::array<std::pair<std::string,
+                              // info::device::type>>
+  auto It = std::find_if(
+      std::begin(DeviceTypeMap), std::end(DeviceTypeMap), [&](auto DtPair) {
+        return std::string::npos != TopDeviceStr.find(DtPair.first);
+      });
+  if (It != DeviceTypeMap.end()) {
+    Target.HasDeviceType = true;
+    Target.DeviceType = It->second;
+    // Handle wildcard.
+    if (TopDeviceStr[0] == '*') {
+      Target.HasDeviceWildCard = true;
+      Target.HasDeviceType = false;
+    }
+  } else { // Only thing left is a number.
+    std::string TDS(TopDeviceStr);
+    try {
+      Target.DeviceNum = std::stoi(TDS);
+      Target.HasDeviceNum = true;
+    } catch (...) {
+      std::stringstream ss;
+      ss << "error parsing device number: " << TDS;
+      throw sycl::exception(sycl::make_error_code(errc::invalid), ss.str());
+    }
+  }
+
+  if (DeviceSubPair.size() == 2) {
+    // We have a subdevice.
+    std::string_view SubDeviceStr = DeviceSubPair[1];
+    // SubDeviceStr is wildcard or number.
+    if (SubDeviceStr[0] == '*') {
+      Target.HasSubDeviceWildCard = true;
+    } else {
+      std::string SDS(SubDeviceStr);
+      try {
+        Target.SubDeviceNum = std::stoi(SDS);
+        Target.HasSubDeviceNum = true;
+      } catch (...) {
+        std::stringstream ss;
+        ss << "error parsing sub-device index: " << SDS;
+        throw sycl::exception(sycl::make_error_code(errc::invalid), ss.str());
+      }
+    }
+  }
```

Comment:
else if (DeviceSubPair.size() > 2) // throw exception as sub-sub-devices are not permitted